### PR TITLE
Added selection of columns functionality

### LINKED ISSUE #6426: Adding the select facility for variables as a partner to the filter system
This has 2 parts.  It builds on, and supersedes topics mentioned in #4372, in #4957 and also in #6263 and in #4009.
The first part is to (probably) replace the current Prepare > Data Frame > Hide/Show Columns with a new Select Columns dialogue.
a) There will still be the Hide Columns (and Restore Hidden Columns) on the right-click menu as described in #4372, together with information on what is hidden at the bottom - just like for rows when there is a filter.
b) The main dialogue will include hide as an option.  It will probably be largely like the main Prepare > Data Frame > Filter dialogue, so the data selector has selects as objects.  You can choose to define a new select, or use an existing one.  One thing you can do with a select is to hide the columns.
c) As with filter, another route to select is from the Data Options button on each main dialogue.  This currently takes you to a version of the Filter dialogue, but just for that data frame.  Now that will change with an additional button, so you can opt for Rows or Columns.

image

It is already here.  Do we keep these tabs - just what we want, or change them to radio buttons?

Currently when you choose Columns it is almost blank.  This would change to be like the dialogue for Rows.
I suggest it can for now be almost an exact copy of this.  
So the data selector is for selections.  Replace Filter by Selection in most of these controls.  Perhaps on the Rows option change Selected Filter Preview to just Filter Preview.  Then on the Columns have Select Preview.

d) Now the Select sub-dialogue.  This is a parallel of the Filter sub-dialogue and many parts are the same.  Here I start the suggestions on the structure of this sub-dialogue:

image

1) It has the usual data selector at the left-hand side - though it is only needed for one of the options.
2) It has the same controls below, i.e. 
image
But the main box has a single entry - just Condition.
And Filter becomes Select, so Select Preview, and Select Name

3) Then the Filter By  becomes Select By: And the single receiver becomes a pull down with the options specified below.
image

The keyboard moves down and may be different.
The Add Condition button remains.
The two controls above change depending on which option is chosen in the pull-down.

The pull-down contains:
a) columns
b) starts_with
c) ends_with
d) contains
e) matches
f) num-range
g) where
h) last_col
i) What else?

These are described in the tidyselect package as well as under select in dplyr.

Now, if it is `columns`, then there is simply a multiple receiver..  Then the Add button is enabled and once one or more columns are selected, the Add Condition is enabled.  The action is then to take the columns and generate c(column1, column2,...) as the condition.
We need to allow for !, i.e. Not.  That could be a checkbox, but I think it could also be added to the list of buttons once the condition is added?  (This has also been suggested for the filter - see issue #6101 

If it is `starts_with`   then the field just receives a string, so the Add Condition becomes `starts_with("string")`There is also a checkbox (default ticked) with `Ignore Case` as the text
Same for `ends_with`, and also `contains` 
(I am not sure about `matches` and `where`.  Where looks very powerful, and there might be some examples given.
If num_range then there are 2 more controls in addition to the string one which is called `Prefix`.  The next is called range.  It us usually something like 10:15.  I haven't checked whether it can be 3,5,8,11, presumably adding c to become c(3,5,8,11).  Then there is a check-box and if checked, then digits can be a pull-down with values 1, 2, 3, 4.  T(this allows for 001, 002, 003, etc if given as 1:3.
 The & and or (|) work just the same as the filters.  

Then selects should be just like filters, so you can rename them, delete them etc.

The details and implementation will benefit from nifty R skills, so am suggesting the @shadrackkibet and @Cal-Insta could perhaps work together on this.

### PR DESCRIPTION
**Pull request by shadrackkibet**
fixes #6426

This is not yet ready for review. I will send a notification when it is ready. 

Design screenshots below.

image
image
image
image

### CODE REVIEW

**Review by rdstern — COMMENTED** [review]
@shadrackkibet great this has started.  I couldn't see where to check?  I assume eventually on Data Options, and a replacement dialogue for Prepare > Data Frame > Hide/Show?

**Comment by shadrackkibet** [comment]
Let us test it from my end at this stage. I will be in touch on skype.

**Comment by shadrackkibet** [comment]
@rdstern hopefully you can test now.

**Comment by rdstern** [comment]
This is great - and seems to work.  Happy to imporve if you think merging at this stage will help.
What's next for you (and for others if you are bringing them in.  Here are some points from this first look:
a) Could we make it easy to have all variables of a given type, e.g. all factors, all logical? 
b) Could it say how many variables are selected.  I tried some that didn't work and got none selected.  I wonder if there should be a warning in this case.
c) I wanted to select all that start with anything and then i in the second place.  I tried .i, but it didn't work.  Could there be a fixed, or regex option in the choice?
d) In the selector the top usually says Variables.  When there is a selection, could there be perhaps simply `(select1)` after this?
e) I look forward to being able to access these from the data options.  The tabs currently say Rows Columns .  I suggest `Rows (Filter) Columns (Select)`
Very exciting!

**Comment by rdstern** [comment]
A job for @dannyparsons on the R side I assume.  I am keen to encourage more use of variable labels in our use of R.  Could @dannyparsons add the feature to be able to search in other attributes as well as the name.  Then, you add a group box in the sub-dialogue It has the label Select from:  Then there are 3 options inside.   Variable name,  Variable Label, Value Label.  Variable Name is ticked by default, and the others are not.

**Comment by rdstern** [comment]
@shadrackkibet could you please also add the help context number 597.  Thanks

**Review by rdstern — APPROVED** [review]
Good to merge if possible now.  Then there are more dimensions of the general facility to add.  
@lloyddewit could you check?

File: instat/dlgSelectColumns.vb [hunk 1]
```diff
@@ -1,3 +1,257 @@
...
+                strValue = clsFromToOperation.ToScript
+                'Case "Where"
+                'clsCurrentConditionList.AddParameter("operation", Chr(34) & "where" & Chr(34), iPosition:=0)
+                'If ucrInputColumnType.GetText = "Numeric" Then
+                'clsParametersList.AddParameter("fn", "is.numeric", iPosition:=0)
+                'ElseIf ucrInputColumnType.GetText = "Factor" Then
+                'clsParametersList.AddParameter("fn", "is.factor", iPosition:=0)
+                'ElseIf ucrInputColumnType.GetText = "Character" Then
+                'clsParametersList.AddParameter("fn", "is.character", iPosition:=0)
+                'End If
```
Comment (lloyddewit) [inline instat/dlgSelectColumns.vb:196]: Delete?

File: instat/dlgSelectColumns.vb [hunk 2]
```diff
@@ -1,3 +1,257 @@
...
+        ucrInputSelectOperation.SetDropDownStyleAsNonEditable()
+
+        ucrInputColumnType.SetItems({"Numeric", "Factor", "Character"})
+        ucrInputColumnType.SetDropDownStyleAsNonEditable()
+
+        ucrInputSelectOperation.AddToLinkedControls(ucrChkIgnoreCase, {"Starts with", "Ends with", "Contains", "Matches"}, bNewLinkedHideIfParameterMissing:=True)
+        ucrInputSelectOperation.AddToLinkedControls(ucrReceiverMultipleVariables, {"Columns"}, bNewLinkedHideIfParameterMissing:=True)
+        ucrInputSelectOperation.AddToLinkedControls({ucrNudFrom, ucrNudTo}, {"Numeric range"}, bNewLinkedAddRemoveParameter:=True, bNewLinkedHideIfParameterMissing:=True)
+        ucrInputSelectOperation.AddToLinkedControls(ucrInputText, {"Numeric range", "Starts with", "Ends with", "Contains", "Matches"}, bNewLinkedHideIfParameterMissing:=True)
+        'ucrInputSelectOperation.AddToLinkedControls(ucrInputColumnType, {"Where"}, bNewLinkedHideIfParameterMissing:=True)
```
Comment (lloyddewit) [inline instat/dlgSelectColumns.vb:57]: delete?

File: instat/dlgSelectColumns.vb [hunk 3]
```diff
@@ -1,3 +1,257 @@
...
+                    bEnableOrDisable = False
+                End If
+            Case "Numeric range"
+                If ucrNudFrom.IsEmpty OrElse ucrNudTo.IsEmpty OrElse ucrInputText.IsEmpty Then
+                    bEnableOrDisable = False
+                End If
+            Case "Where"
+                '
+            Case "Last column"
+                '
```
Comment (lloyddewit) [inline instat/dlgSelectColumns.vb:236]: delete?

File: instat/dlgSelectColumns.vb [hunk 4]
```diff
@@ -1,3 +1,257 @@
...
+        cmdAddCondition.Enabled = bEnableOrDisable
+    End Sub
+
+    Private Sub ucrReceiverMultipleVariables_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrReceiverMultipleVariables.ControlValueChanged, ucrInputSelectOperation.ControlValueChanged, ucrNudFrom.ControlValueChanged, ucrNudTo.ControlValueChanged
+        EnableDisableAddConditionButton()
+    End Sub
+
+    Private Sub ucrInputText_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrInputText.ControlContentsChanged
+        EnableDisableAddConditionButton()
+    End Sub
```
Comment (lloyddewit) [inline instat/dlgSelectColumns.vb:247]: combine these into a single handler?

**Review by lloyddewit — CHANGES_REQUESTED** [review]
This looks good, I just had some small comments

File: instat/dlgSelectColumns.vb (hunk 1 above)
Comment (shadrackkibet) [reply instat/dlgSelectColumns.vb:196]: I need it when we add functionality to select factors, characters etc columns. I will add a comment!

File: instat/dlgSelectColumns.vb (hunk 2 above)
Comment (shadrackkibet) [reply instat/dlgSelectColumns.vb:57]: For the same reason as above

File: instat/dlgSelectColumns.vb (hunk 3 above)
Comment (shadrackkibet) [reply instat/dlgSelectColumns.vb:236]: same case

File: instat/dlgSelectColumns.vb (hunk 4 above)
Comment (shadrackkibet) [reply instat/dlgSelectColumns.vb:247]: agreed

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@rdstern Since your last approval, there are only small changes that should not affect functionality. Please could you approve again?
We can then merge, thanks

**Comment by shadrackkibet** [comment]
@lloyddewit many thanks for the reviews.

**Comment by shadrackkibet** [comment]
@rdstern?

**Review by rdstern — APPROVED** [review]
Great to merge, so I am approving.  There are many things to improve of course - and I just noticed another one - maybe two!
a) The default iin the sub-dialogue s to select the columns and that is fine.  There is still the Data Options button, and it is active.  I am not sure it should be there!
b) When I move to the Starts, etc, then the column selector remains visible and I like that.  I'd like even to be able to scroll there?  But we should disable the Add button to indicate that the list of variables is then just for viewing.

**Pull request merged**